### PR TITLE
Update clipboard Qt-bindings for flexiblity and Python3 compatibility

### LINKED ISSUE #17722: Qtbindings clipboard conflict
In the clipboard package, I see pandas is using pyperclip

the problem is that pyperclip uses PyQt4 when deciding to use Qt for the linux clipboard, which can create a conflict if trying to embed pandas into a PyQt5 or PySide application for example

I think the simplest solution is to use [qtpy from the Spyder](https://github.com/spyder-ide/qtpy) team, so replace:

```python
from PyQt4.QtGui import QApplication
```

with

```python
from qtpy.QtWidgets import QApplication
```

I am currently working on a PyQt5 based pandas data visualization program and this is a problem I have encountered in the past.

### PR DESCRIPTION
**Pull request by dvincentwest**
This should prevent any conflicts with other qt-bindings packages when embedding pandas in a Qt-based gui and should also provide compatibility with Python3 since PyQt4 may not be available for the latest releases.

- [ ] closes #17722

### CODE REVIEW

File: pandas/io/clipboard/__init__.py [hunk 1]
```diff
@@ -68,8 +69,8 @@ def determine_clipboard():
             return init_gtk_clipboard()
 
         try:
-            # Check if PyQt4 is installed
-            import PyQt4  # noqa
+            # Check if qtpy is installed
+            import qtpy  # noqa
```
Comment (TomAugspurger) [inline pandas/io/clipboard/__init__.py]: Shouldn't this fall back to `PyQt4` if `qtpy` isn't available? I'm Ok with making `qtpy` the default, but we don't want to break a user's install that just has PyQt4.

**Review by TomAugspurger — COMMENTED** [review]
You'll also want to update `doc/source/install.rst` under Optional Dependencies

**Comment by sinhrks** [comment]
Should update `io.rst` too.

**Comment by dvincentwest** [comment]
I will make all these updates within a couple of days

File: pandas/io/clipboard/clipboards.py [hunk 2]
```diff
@@ -46,20 +46,24 @@ def paste_gtk():
...
+
+    app = QApplication.instance()
+    if app is None:
+        app = QApplication([])
 
     def copy_qt(text):
         cb = app.clipboard()
         cb.setText(text)
 
-    def paste_qt():
```
Comment (jreback) [inline pandas/io/clipboard/clipboards.py]: why did you remove this?

**Comment by dvincentwest** [comment]
@asweigart just pushed some commits to pyperclip, more than I was tackling here, but I reverted the code to his latest commits regarding this issue.  Behavior now is to import qtpy, then PyQt5 if that fails, then PyQt4 if that fails.  Should not break for anyone who was using this functionality previously.  Documentation was also updated as requested.  I am waiting for CI checks to pass, and will check back to make sure that they do

**Comment by dvincentwest** [comment]
switched to vim key-binding recently, getting used to it, but it throws me
off once in awhile.  I added it back

**Comment by dvincentwest** [comment]
it appears I violated style standards with 2 greater than 80 character lines.  I should have read the contribution documentation more carefully.  Apologies, I will review that, fix the lines, and update.

**Comment by jreback** [comment]
ideally would like to enhance 1 or more of our CI builds to use qtpy. Some of them are already testing  via xsel/python-gtk2. they are a bit flaky though. can you add in pyqt to a couple of builds (that don't otherwise add the x selections).  the clipboard tests themselves exercise this.

**Comment by dvincentwest** [comment]
@jreback, I've never done that, but in the spirit of learning I will try, but will be careful before making any changes.  On a second note, the last round of tests failed because ModuleNotFoundError was not recognized as a valid exception.  Why is that?  I changed the exception to be ImportError instead, which is valid and should work, but I was wondering why the error with ModuleNotFoundError.

**Comment by jreback** [comment]
> ModuleNotFoundError
is python3 only. ImportError will be raised for py2

**Comment by ccordoba12** [comment]
I'm coming here from spyder-ide/spyder#5795. This problem is also generating odd errors for us in Spyder, so I want to chime in to see when this is going to be merged and if there's something I can do to help with.

@jreback and @TomAugspurger, I think Pandas should depend directly on `qtpy`. `qtpy` is a tiny dependency and it would make the code added by @dvincentwest (which looks right to me) very, very simple, instead of a lot of try/excepts (which are already done in `qtpy` itself).

Besides, it would be easier to handle future renames or changes in Qt by depending on `qtpy` because `qtpy` will take care of them.

**Comment by TomAugspurger** [comment]
Thanks.

I don't think adding qtpy as a dependency is worthwhile, since `read_clipiboard` isn't a core part of pandas.

@dvincentwest could you add a release note? I think in `doc/source/whatsnew/v0.22.0.txt`. Other than that, the changes here look good.

File: doc/source/install.rst [hunk 3]
```diff
@@ -251,7 +251,8 @@ Optional Dependencies
 * `Jinja2 <http://jinja.pocoo.org/>`__: Template engine for conditional HTML formatting.
 * `s3fs <http://s3fs.readthedocs.io/>`__: necessary for Amazon S3 access (s3fs >= 0.0.7).
 * `blosc <https://pypi.python.org/pypi/blosc>`__: for msgpack compression using ``blosc``
-* One of `PyQt4
```
Comment (jreback) [inline doc/source/install.rst]: you can list qtpy here as well

File: doc/source/whatsnew/v0.22.0.txt [hunk 4]
```diff
@@ -127,6 +127,7 @@ I/O
 - Bug in :func:`read_msgpack` with a non existent file is passed in Python 2 (:issue:`15296`)
 - Bug in :func:`read_csv` where a ``MultiIndex`` with duplicate columns was not being mangled appropriately (:issue:`18062`)
 - Bug in :func:`read_sas` where a file with 0 variables gave an ``AttributeError`` incorrectly. Now it gives an ``EmptyDataError`` (:issue:`18184`)
+- :func:`pandas.read_clipboard` updated to use qtpy, falling back to PyQt5 and then PyQt4, adding compatibility with Python3 and multiple python-qt bindings (:issue:`17722`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: move to other enhancements

**Review by jreback — CHANGES_REQUESTED** [review]
you need to add qtpy to the ci/requirements-3.6_BUILD_TEST.sh

**Comment by jreback** [comment]
does installing ``qtpy`` actually installing a pyqt binding? e.g the ``qt`` package?

**Comment by dvincentwest** [comment]
No, qtpy is only a compatibility later so that you can write bindings agnostic applications. It still requires either Pyqt(4 or 5) or Pyside(1 or 2)

**Comment by jreback** [comment]
> No, qtpy is only a compatibility later so that you can write bindings agnostic applications. It still requires either Pyqt(4 or 5) or Pyside(1 or 2)

ok can you make that clear in the install.rst instructions. also do add to one of the ci builds.

**Comment by dvincentwest** [comment]
yes, I will try to get this done tomorrow...

**Comment by dvincentwest** [comment]
@jreback @TomAugspurger , the Python 2.7 appveyor build appears to have gone over the 1 hour time-limit.  I am not sure why, and I don't really know what to do about that next.

I did add pyqt (conda label for PyQt5) and qtpy to the `ci/requirements-3.6_BUILD_TEST.sh` script but I have never dealt with CI before so I think you should have a quick look to make sure I did what you were looking for.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @dvincentwest

**Comment by dvincentwest** [comment]
And my thanks to you guys for maintaining the project

**Pull request merged**